include/circt/Query/Query.h

```diff
@@ -0,0 +1,217 @@
+#ifndef CIRCT_QUERY_API_H
+#define CIRCT_QUERY_API_H
+
+#include <iostream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "circt/Dialect/Comb/CombDialect.h"
+#include "circt/Dialect/Comb/CombVisitors.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWTypes.h"
+#include "circt/Dialect/HW/HWVisitors.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/SV/SVVisitors.h"
+#include "llvm/Support/Regex.h"
+
+namespace circt {
+namespace query {
+
+class FilterType {
+public:
+  virtual ~FilterType() {}
+
+  virtual bool valueMatches(llvm::StringRef value) = 0;
+  virtual bool addSelf() { return false; }
+  virtual FilterType *clone() = 0;
+
+protected:
+  FilterType() {}
+};
+
+class EmptyFilterType : public FilterType {
+public:
+  virtual bool valueMatches(llvm::StringRef value) override { return false; }
+  virtual FilterType *clone() override { return new EmptyFilterType; }
+};
+
+class GlobFilterType : public FilterType {
+public:
+  GlobFilterType() {}
+
+  virtual bool valueMatches(llvm::StringRef value) override { return true; }
+  virtual FilterType *clone() override { return new GlobFilterType; }
+};
+
+class RecursiveGlobFilterType : public GlobFilterType {
+public:
+  RecursiveGlobFilterType() {}
+
+  virtual bool addSelf() override { return true; }
+  virtual FilterType *clone() override { return new RecursiveGlobFilterType; }
+};
+
+class LiteralFilterType : public FilterType {
+public:
+  LiteralFilterType(std::string &literal) : literal(literal) {}
+
+  virtual bool valueMatches(llvm::StringRef value) override {
+    return value == literal;
+  }
+  virtual FilterType *clone() override {
+    return new LiteralFilterType(literal);
+  }
+
+private:
+  std::string literal;
+};
+
+class RegexFilterType : public FilterType {
+public:
+  RegexFilterType(llvm::StringRef regex)
+      : repr(regex.str()), regex(llvm::Regex(regex)) {}
+
+  virtual bool valueMatches(llvm::StringRef value) override {
+    return regex.match(value);
+  }
+  virtual FilterType *clone() override { return new RegexFilterType(repr); }
+
+private:
+  std::string repr;
+  llvm::Regex regex;
+};
+
+class FilterData {
+public:
+  FilterData(Operation *root)
+      : tables(mlir::SymbolTableCollection()),
+        userMap(mlir::SymbolUserMap(tables, root)) {}
+
+  mlir::SymbolTableCollection tables;
+  mlir::SymbolUserMap userMap;
+};
+
+class Filter {
+public:
+  virtual ~Filter() {}
+
+  virtual bool matches(Operation *op, FilterData &data) = 0;
+  virtual bool addSelf() { return type->addSelf(); }
+  virtual Filter *nextFilter() { return nullptr; }
+  virtual Filter *clone() = 0;
+  virtual std::vector<Operation *> nextOperations(Operation *op,
```

Comment:
nextOperations seems an odd name.  Please add comments.
What is this used for?  It seems odd to define a method which just pushes into a vector.
Also, SmallVector can be returned by value.